### DISCUSSION TITLE
Is there any API to fetch all nodes/tags  present in the OPCUA server.

### DISCUSSION THREAD

**Discussion by sumit-quartic** [FreeOpcUa/opcua-asyncio#645] [Q&A]
Hi,

My requirement is to fetch all the node/stags present in the OPCUA server. For that,  currently I am traversing recursively using the following code. However, I wanted to know if there is any API to do the same.  

async def browse_recursive(self, node):
        nodes = await node.get_children()
        for childId in nodes:
            ch = self._opcua_client.get_node(childId)
            node_class = await ch.read_node_class()           
            if node_class == ua.NodeClass.Object:                
                await self.browse_recursive(ch)
            elif node_class == ua.NodeClass.Variable:              
                try:
                    bn = await ch.read_browse_name()
                    dn = await ch.read_display_name()
                    val= str( await ch.get_value()) 
                    print(str(childId.nodeid))
                except Exception as e:
                    print (e)
                    pass
            else:               
                pass

**Comment by swamper123** [comment]
There is a method called: `browse_nodes(nodes)` for the client, so you browse them all in one call, instead of flooding your network for each one.
`get_children()` returns only hierarchical Nodes, if I got that right in my mind. Saying, nodes with nonhierarchical references may not appear in the list.

But are you realy sure you want to fetch ALL nodes with ALL Attributes they have of the server? Even all types, server specific nodes etc? Or just the ones you need? Do you want to aggregate a server?

Btw. I found in the examples a small strip of code, you might be interessted in:

https://github.com/FreeOpcUa/opcua-asyncio/blob/84b4a60f39c9d71e4c22b9c036611783d03de13d/examples/client-minimal-auth.py

It's not exactly what you are looking for, but you may use some lines of that.

**Reply by sumit-quartic** [reply]
Thanks for replying @swamper123 , Input to my method will be either a node_id or name space such as "ns=2", now I want to fetch the node ids of all nodes which are in the input node tree or the nodes which are under name space "2". My requirement is to fetch the nodes of type "variable" only.

**Reply by swamper123** [reply]
Hmm...I just found a method on server side:

https://github.com/FreeOpcUa/opcua-asyncio/blob/854c2e4962eee9aa45bcd5f29df32afe6618baa9/asyncua/common/ua_utils.py#L256

But it seems, that you have to do it on your own. Perhaps the link to the first example may help you out and has to be modified.

It would be cool, if you would make a PR for that (best place would be client or asyncua/common).
